monix-eval/shared/src/main/scala/monix/eval/internal/TaskTracedRunLoop.scala

```diff
@@ -0,0 +1,475 @@
+    ArrayStack(8)
+
+  /** Internal utility, for forcing an asynchronous boundary in the
+    * trampoline loop.
+    */
+  def restartAsync[A](
+                       source: Task[A],
+                       context: Context,
+                       cb: Callback[A],
+                       bindCurrent: Bind,
+                       bindRest: CallStack,
+                       local: Local.Context): Unit = {
+    val oldCtx = Local.getContext()
+    if (!context.shouldCancel)
+      context.scheduler.executeAsync { () =>
+        // Resetting the frameRef, as a real asynchronous boundary happened
+        context.frameRef.reset()
+        startWithCallback(source, context, cb, bindCurrent, bindRest, 1, local)
+      }
+  }
+
+  /** Logic for finding the next `Transformation` reference,
+    * meant for handling errors in the run-loop.
+    */
+  private def findErrorHandler(bFirst: Bind, bRest: CallStack): Transformation[Any, Task[Any]] = {
+    var result: Transformation[Any, Task[Any]] = null
+    var cursor = bFirst
+    var continue = true
+
+    while (continue) {
+      if (cursor != null && cursor.isInstanceOf[Transformation[_, _]]) {
+        result = cursor.asInstanceOf[Transformation[Any, Task[Any]]]
+        continue = false
+      } else {
+        cursor = if (bRest ne null) bRest.pop() else null
+        continue = cursor != null
+      }
+    }
+    result
+  }
+
+
+  private def popNextBind(bFirst: Bind, bRest: CallStack): Bind = {
+    if ((bFirst ne null) && !bFirst.isInstanceOf[Transformation.OnError[_]])
+      bFirst
+    else if (bRest ne null) {
+      var cursor: Bind = null
+      do { cursor = bRest.pop() }
+      while (cursor != null && cursor.isInstanceOf[Transformation.OnError[_]])
+      cursor
+    } else {
+      null
+    }
+  }
+
+  /** Internal utility, starts or resumes evaluation of
+    * the run-loop from where it left off.
+    *
+    * The `frameIndex=1` default value ensures that the
+    * first cycle of the trampoline gets executed regardless of
+    * the `ExecutionModel`.
+    */
+  def startWithCallback[A](
+                            source: Task[A],
+                            context: Context,
+                            cb: Callback[A],
+                            bindCurrent: Bind,
+                            bindRest: CallStack,
+                            frameIndex: FrameIndex,
+                            local: Local.Context): Unit = {
+
+    final class RestartCallback(context: Context, callback: Callback[Any]) extends Callback[Any] {
+      private[this] var canCall = false
+      private[this] var bFirst: Bind = _
+      private[this] var bRest: CallStack = _
+      private[this] val runLoopIndex = context.frameRef
+
+      def prepare(bindCurrent: Bind, bindRest: CallStack): Unit = {
+        canCall = true
+        this.bFirst = bindCurrent
+        this.bRest = bindRest
+      }
+
+      def onSuccess(value: Any): Unit =
+        Local.withContext(local) {
```

Comment:
Why use `Local.withContext` here instead of `Local.setContext`? 

Can you give me an example of why restoring that context at the end of the block matters?

Reply:
@alexandru testing the CorrelationIdSuite after making this change:

```scala
      def onSuccess(value: Any): Unit = {
        Local.setContext(local)
          // This is where finally the local gets propagated when executing
          if (canCall) {
            canCall = false
            loop(Now(value), context.executionModel, callback, this, bFirst, bRest, runLoopIndex())
          }
        }
```

Throws errors on assertions because in the execution a thread might get reused so its context may still exist and tests like this might fail:

```scala
test("should NOT get CorrelarionId with flatmapped Task with async boundary") {
    // Does not work without TracingScheduler
    // Meaning that when mapping an executed Task, the future will always
    // need the TracingScheduler to propagate the CorrelationId

    val t1 = CorrelationId("0000").asCurrent {
      import monix.execution.Scheduler.Implicits.global
      taskFlatMap.runAsyncTraced.map {
        case (x, v) =>
          (x, CorrelationId.current)
      }
    }
    val (_, cid1) = Await.result(t1, 10.seconds)
    assert(cid1.isEmpty)

    val t2 = CorrelationId("0000").asCurrent {
      import monix.execution.Scheduler.Implicits.global
      taskFlatMap.runAsync.map {
        case (x, v) =>
          (x, CorrelationId.current)
      }
    }
    val (_, cid2) = Await.result(t2, 10.seconds)
    assert(cid2.isEmpty)
  }
```

Because the Local value is still defined in that thread. I reference this comment [Source](http://stevenskelton.ca/threadlocal-variables-scala-futures/) :
```
A small note about garbage collection. As long as a thread exists, it will maintain references
to its ThreadLocal variables. If the thread-pool does not recycle threads and a thread goes
back into the pool without releasing its TLS then those objects will not be freed
```

At least for tracing contexts, we need to use only the current local and restore the previous state so we don't have leaked contexts.

Reply:
I see, OK.